idx/memory/memory.go

```diff
@@ -0,0 +1,475 @@
+package memory
+
+import (
+	"flag"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/raintank/met"
+	"github.com/raintank/metrictank/idx"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+var (
+	idxOk             met.Count
+	idxFail           met.Count
+	idxAddDuration    met.Timer
+	idxGetDuration    met.Timer
+	idxListDuration   met.Timer
+	idxFindDuration   met.Timer
+	idxDeleteDuration met.Timer
+
+	Enabled bool
+)
+
+func ConfigSetup() {
+	memoryIdx := flag.NewFlagSet("memory-idx", flag.ExitOnError)
+	memoryIdx.BoolVar(&Enabled, "enabled", false, "")
+	globalconf.Register("memory-idx", memoryIdx)
+}
+
+type Tree struct {
+	Items map[string]*Node
+}
+
+type Node struct {
+	Path     string
+	Children []string
+	Leaf     bool
+}
+
+func (n *Node) String() string {
+	t := "branch"
+	if n.Leaf {
+		t = "leaf"
+	}
+	return fmt.Sprintf("%s - %s", t, n.Path)
+}
+
+// Implements the the "MetricIndex" interface
+type MemoryIdx struct {
+	sync.Mutex
+	DefById map[string]*schema.MetricDefinition
+	Tree    map[int]*Tree
```

Comment:
We tend to overuse pointers in various places.
e.g. AggMetrics has a `map[string]*AggMetric`, while `map[string]AggMetric` would suffice, since whenever we want to maintain modifications to AggMetric's we can (and already do) use methods with pointer receivers and/or functions with pointer arguments. So we only need pointers to the structures very temporarily (while they're being passed into functions/methods), not at all times.

storing big lists or maps of pointers (or even a lot of strings, since strings internally also have references/pointers to them):
1) exerts needless strain on the garbage collector, since it has to scan all the live pointers at every GC run.
2) leads to memory fragmentation
3) consumes more memory than needed (i believe 8B per pointer on a 64bit system)

this is why:
1) the old idx package used plain integers to represent values (a trick tought by dgryski)
2) i want to update a few key structures such as AggMetrics to use less pointers and hopefully see a measurable effect. supposedly, it should make our GC issues less severe.
3) we can hopefully transition to use less and less pointers in general.